plugins/inputs/azure_storage_queue/azure_storage_queue.go

```diff
@@ -19,108 +19,101 @@ import (
+	if err != nil {
+		return fmt.Errorf("creating client failed: %w", err)
+	}
+	a.client = client
+
 	return nil
 }
 
 func (a *AzureStorageQueue) Gather(acc telegraf.Accumulator) error {
-	serviceURL, err := a.getServiceURL()
-	if err != nil {
-		return err
-	}
+	ctx := context.Background()
 
-	ctx := context.TODO()
+	a.Log.Debugf("Listing queues of storage account %q", a.StorageAccountName)
 
-	for marker := (azqueue.Marker{}); marker.NotDone(); {
-		a.Log.Debugf("Listing queues of storage account %q", a.StorageAccountName)
-		queuesSegment, err := serviceURL.ListQueuesSegment(ctx, marker,
-			azqueue.ListQueuesSegmentOptions{
-				Detail: azqueue.ListQueuesSegmentDetails{Metadata: false},
-			})
+	// Iterate through the queues and generate metrics
+	pages := a.client.NewListQueuesPager(nil)
+	for pages.More() {
+		response, err := pages.NextPage(ctx)
 		if err != nil {
-			return err
+			return fmt.Errorf("getting next page failed: %w", err)
 		}
-		marker = queuesSegment.NextMarker
 
-		for _, queueItem := range queuesSegment.QueueItems {
-			a.Log.Debugf("Processing queue %q of storage account %q", queueItem.Name, a.StorageAccountName)
-			queueURL := serviceURL.NewQueueURL(queueItem.Name)
-			properties, err := queueURL.GetProperties(ctx)
+		// Get the properties and the message properties for each of the queues
+		for _, queue := range response.Queues {
+			if queue.Name == nil {
+				continue
+			}
+			name := strings.TrimSpace(*queue.Name)
+
+			// Access the queue and get the properties
+			c := a.client.NewQueueClient(*queue.Name)
+			props, err := c.GetProperties(ctx, nil)
 			if err != nil {
-				a.Log.Errorf("Error getting properties for queue %s: %s", queueItem.Name, err.Error())
+				acc.AddError(fmt.Errorf("getting properties for queue %q failed: %w", name, err))
 				continue
 			}
-			var peekedMessage *azqueue.PeekedMessage
+			if props.ApproximateMessagesCount == nil {
+				acc.AddError(fmt.Errorf("unset message count for queue %q", name))
+				continue
+			}
+
+			// Setup the metric elements
+			tags := map[string]string{
+				"account": a.StorageAccountName,
+				"queue":   strings.TrimSpace(name),
+			}
+			fields := map[string]interface{}{
+				"size": *props.ApproximateMessagesCount,
+			}
+			now := time.Now()
 			if a.PeekOldestMessageAge {
-				messagesURL := queueURL.NewMessagesURL()
-				messagesResponse, err := messagesURL.Peek(ctx, 1)
-				if err != nil {
-					a.Log.Errorf("Error peeking queue %s: %s", queueItem.Name, err.Error())
-				} else if messagesResponse.NumMessages() > 0 {
-					peekedMessage = messagesResponse.Message(0)
+				if r, err := c.PeekMessage(ctx, nil); err != nil {
+					acc.AddError(fmt.Errorf("peeking message for queue %q failed: %w", name, err))
+				} else if len(r.Messages) > 0 && r.Messages[0] != nil && r.Messages[0].InsertionTime != nil {
+					msg := r.Messages[0]
+					fields["oldest_message_age_ns"] = now.Sub(*msg.InsertionTime).Nanoseconds()
```

Comment:
`r.Messages[0] != nil` is unnecessary since r.Messages[0] is not a pointer.
Instead of checking InsertionTime inline, how about extract the message and handle nil cases separately

```suggestion
				} else if len(r.Messages) > 0 {
                                            msg := r.Messages[0]
                                            if msg.InsertionTime != nil {
                                                fields["oldest_message_age_ns"] = now.Sub(*msg.InsertionTime).Nanoseconds()
                                            }
                    ```

Reply:
Hmmm why do you think so? The variable `r` is filled using [`QueueClient.PeekMessage()`](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/storage/azqueue#QueueClient.PeekMessage) returning a `PeekMessagesResponse` which corresponds to a [`generated.MessagesClientPeekResponse`](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/storage/azqueue@v1.0.0/internal/generated#MessagesClientPeekResponse) type. If you check this type then `Messages` is defined as [`Messages []*PeekedMessage`](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/storage/azqueue@v1.0.0/internal/generated#PeekedMessage), i.e. a slice of pointers to the peeked messages. If you now access `r.Messages[0]` you get an instance of type `*PeekedMessage` and need to check if that pointer is `nil`. Am I missing something here?

Reply:
You're absolutely right! It is indeed a slice of pointers to the peeked messages. I should have used my IDE for reviewing, as it makes it easier to check types and follow through. My mistake, closing this comment.